compactor: purge suggestions that have live data

benesch opened the pull request:
While debugging #24029 we discovered that RESTORE generates massive
numbers of suggested compactions as it splits and scatters ranges. As
the cluster rebalances, every removed replica leaves behind a range
deletion tombstone and a suggested compaction over the keys it covered.

Occasionally, the replica chosen for rebalancing will be a member of the
last range of the cluster. This range extends from wherever the restore
has last split the table it's restoring to the very last key. Suppose
we're restoring a table with 10,000 ranges evenly distributed across
primary keys from 1-10,000. If a replica in the last range gets
rebalanced early in the restore—say, after only the first 500 ranges
have been split off—at least one node in the cluster will have a
suggested compaction for a range like the following:

    /Table/51/1/500 - /Max

This creates a huge problem! The restore will eventually create 9500
more ranges in that keyspan, each about 32MiB in size. Some of those
ranges will necessarily rebalance back onto the node with the suggested
compaction. By the time the compaction queue gets around to processing
the suggestion, there might be hundreds of gigabytes within the range.

In our 2TB (replicated) store dump, snapshotted immediately after a
RESTORE, there were two such massive suggested compactions, each of
which took over 1h to complete. This bogs down the compaction queue with
unnecessary work and makes it especially dangerous to initiate a DROP
TABLE (#24029), as the incoming range deletion tombstones will pile up
until the prior compaction finishes, and the cluster grinds to a halt in
the meantime.

The same problem happens whenever any replica is rebalanced away and
back before the compaction queue has a chance to compact away the range
deletion tombstone, though the impact is limited because the keyspan is
smaller.

This commit prevents the compaction queue from getting bogged down with
suggestions based on outdated information. At the time the suggestion is
considered for compaction, the queue checks whether the key span
suggested has any live keys. The sign of even a single key is a good
indicator that something has changed—usually that the replica, or one of
its split children, has been rebalanced back onto the node. The
compaction queue now deletes this suggestion instead of acting on it.

This is a crucial piece of the fix to #24029. The other half, #26449,
involves rate-limiting ClearRange requests.

I suspect this change will have a nice performance boost for RESTORE.
Anecdotally we've noticed that restores slow down over time. I'm willing
to bet its because nonsense suggested compactions start hogging disk
I/O.

Release note: None

benesch committed:
compactor: purge suggestions that have live data

While debugging #24029 we discovered that RESTORE generates massive
numbers of suggested compactions as it splits and scatters ranges. As
the cluster rebalances, every removed replica leaves behind a range
deletion tombstone and a suggested compaction over the keys it covered.

Occasionally, the replica chosen for rebalancing will be a member of the
last range of the cluster. This range extends from wherever the restore
has last split the table it's restoring to the very last key. Suppose
we're restoring a table with 10,000 ranges evenly distributed across
primary keys from 1-10,000. If a replica in the last range gets
rebalanced early in the restore—say, after only the first 500 ranges
have been split off—at least one node in the cluster will have a
suggested compaction for a range like the following:

    /Table/51/1/500 - /Max

This creates a huge problem! The restore will eventually create 9500
more ranges in that keyspan, each about 32MiB in size. Some of those
ranges will necessarily rebalance back onto the node with the suggested
compaction. By the time the compaction queue gets around to processing
the suggestion, there might be hundreds of gigabytes within the range.

In our 2TB (replicated) store dump, snapshotted immediately after a
RESTORE, there were two such massive suggested compactions, each of
which took over 1h to complete. This bogs down the compaction queue with
unnecessary work and makes it especially dangerous to initiate a DROP
TABLE (#24029), as the incoming range deletion tombstones will pile up
until the prior compaction finishes, and the cluster grinds to a halt in
the meantime.

The same problem happens whenever any replica is rebalanced away and
back before the compaction queue has a chance to compact away the range
deletion tombstone, though the impact is limited because the keyspan is
smaller.

This commit prevents the compaction queue from getting bogged down with
suggestions based on outdated information. At the time the suggestion is
considered for compaction, the queue checks whether the key span
suggested has any live keys. The sign of even a single key is a good
indicator that something has changed—usually that the replica, or one of
its split children, has been rebalanced back onto the node. The
compaction queue now deletes this suggestion instead of acting on it.

This is a crucial piece of the fix to #24029. The other half, #26449,
involves rate-limiting ClearRange requests.

I suspect this change will have a nice performance boost for RESTORE.
Anecdotally we've noticed that restores slow down over time. I'm willing
to bet its because nonsense suggested compactions start hogging disk
I/O.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/26488)
<!-- Reviewable:end -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 2 files reviewed at latest revision, all discussions resolved, some commit checks pending.

---

*[pkg/storage/compactor/compactor.go, line 286 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26488#-LELx1r79QTST3SrNB7P:-LELx1r79QTST3SrNB7Q:b-gaoa5r) ([raw file](https://github.com/cockroachdb/cockroach/blob/184cecb8feadec8f4ac205d633bcdd04b22cc870/pkg/storage/compactor/compactor.go#L286)):*
> ```Go
> 
> // fetchSuggestions loads the persisted suggested compactions from the store.
> func (c *Compactor) fetchSuggestions(ctx context.Context) (
> ```

There is duplication between this method and `examineQueue`. Worth getting rid of? The duplication worries me less than the differing behavior. Should `examineQueue` clear unnecessary suggestions?

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26488#-:-LELxPm_E0NlGPYAbv5U:bnfp4nl)*
<!-- Sent from Reviewable.io -->

benesch commented:
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[pkg/storage/compactor/compactor.go, line 286 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26488#-LELx1r79QTST3SrNB7P:-LEM-vgzEayAYbaMxWaU:bf23g68) ([raw file](https://github.com/cockroachdb/cockroach/blob/184cecb8feadec8f4ac205d633bcdd04b22cc870/pkg/storage/compactor/compactor.go#L286)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

There is duplication between this method and `examineQueue`. Worth getting rid of? The duplication worries me less than the differing behavior. Should `examineQueue` clear unnecessary suggestions?
</blockquote></details>

That's a great point. I think I'd prefer to leave as is for the moment if we're planning to back port.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26488)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26488#-:-LEM258U7XIdq6RdgCIG:bnfp4nl)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[pkg/storage/compactor/compactor.go, line 286 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26488#-LELx1r79QTST3SrNB7P:-LEM3xdG5vkONUO11cPs:b1r875t) ([raw file](https://github.com/cockroachdb/cockroach/blob/184cecb8feadec8f4ac205d633bcdd04b22cc870/pkg/storage/compactor/compactor.go#L286)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

That's a great point. I think I'd prefer to leave as is for the moment if we're planning to back port.
</blockquote></details>

Ack. Let's leave as-is.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26488)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

A mild worry that I have is that for some reason the nonempty suggestion case is more common than we think, and that consequently we're compacting much less than we should in practice without having good visibility. Perhaps you can `log.Info` in this case? That's a bit more lightweight than adding a new metric (which may be a pain to backport).

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26488#-:-LEM5d5s53lw8Kt6a42g:b1ch5uh)*
<!-- Sent from Reviewable.io -->

benesch commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[pkg/storage/compactor/compactor.go, line 325 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26488#-LEM6k4m7C2DktvXO2gU:-LEM6k4n4rRGA21BE0x0:b-oovxvv) ([raw file](https://github.com/cockroachdb/cockroach/blob/184cecb8feadec8f4ac205d633bcdd04b22cc870/pkg/storage/compactor/compactor.go#L325)):*
> ```Go
> 				delBatch.Clear(kv.Key)
> 				log.VEventf(ctx, 2, "purging suggested compaction for range %s - %s that contains data",
> 					sc.StartKey, sc.EndKey)
> ```

@tschottdorf want me to just upgrade this to an unconditional Infof, or did you have something else in mind?

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26488)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/storage/compactor/compactor.go, line 325 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26488#-LEM6k4m7C2DktvXO2gU:-LEM7AMgA3jjEu3EIYyc:b-a29ge4) ([raw file](https://github.com/cockroachdb/cockroach/blob/184cecb8feadec8f4ac205d633bcdd04b22cc870/pkg/storage/compactor/compactor.go#L325)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

@tschottdorf want me to just upgrade this to an unconditional Infof, or did you have something else in mind?
</blockquote></details>

Yep. This couldn't happen any more frequently than two change replicas on average, right?

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26488)*
<!-- Sent from Reviewable.io -->

benesch commented:
For posterity: this is a pessimization until #26872 goes in. Waiting on that to land.

benesch committed:
engine: allow clients to update an iterator's upper bound

RocksDB iterators support taking on a new upper bound after creation. We
use this support internally in the engine package to enable a batch's
cached iterator to back many logical iterators with different upper
bounds over its lifetime.

Expose this support to clients outside of the engine package by
introducing a new method, Iterator.SetUpperBound. This method is useful
when a number of separate bounded seeks need to be performed and the
client does not want to create a new iterator for each seek with the
appropriate upper bound.

Release note: None

benesch commented:
This needs a quick re-review. I had to introduce a new API, engine.Iterator.SetUpperBound, to be able to efficiently reuse one iterator when checking for outdated compaction suggestions.

bdarnell reviewed (approved):
> Reviewed 2 of 4 files at r2, 2 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26488)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 3 stale)



<!-- Sent from Reviewable.io -->

tbg reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 4 of 4 files at r2, 2 of 2 files at r3, 1 of 1 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26488#-:-LGzb3dG0BqzXPfJ_mgd:bnfp4nl)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 3 stale)



<!-- Sent from Reviewable.io -->

benesch commented:
bors r=tschottdorf,bdarnell

benesch commented:
Unrelated build failure.

bors r=tschottdorf,bdarnell
